boards: arm: pinetime: Add display backlight to dts

caspermeijn opened the pull request:
Fixes #41047. Define the display backlight hardware relation in the PineTime dts and enable the backlight in the st7789v driver.

1. Add a backlight-led attribute to the display-controller binding. This is for panels with an external backlight led.
2. Add the backlight reference from the st7789v display to the bl led.
3. Make sure led driver is initialized before the display
4. Extend st7789v driver with backlight
5. Enable led driver for pinetime

caspermeijn committed:
dts: bindings: Add backlight-led to st7789v

Add a backlight-led attribute to a led node. This is for
panels with an external backlight led.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
boards: arm: pinetime: Add display backlight to dts

Add the backlight reference from the st7789v display to the bl led.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
display: Init led driver before display driver

Make sure the led driver is initialized before display driver. This is
needed for backlights that are driven by a led driver.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn committed:
display: st7789v: Add led-backlight brightness control

Add ST7789v driver support for led-backlight brightness control.

Signed-off-by: Casper Meijn <casper@meijn.net>

gmarull reviewed (commented):
Have you considered using the regulator API for this purpose?

caspermeijn commented:
> Have you considered using the regulator API for this purpose?

I believe this is different. For now the backlight is just a simple GPIO, but I want to work towards doing PWM to control the backlight level. Therefore I seems logical to use the led API.

For context, I want to extend the display driver to control the backlight. The application can then use `display_set_brightness(const struct device *dev, uint8_t brightness)`.

henrikbrixandersen commented:
If you intend to handle it via the LED API, why is a new binding needed?

caspermeijn commented:
> If you intend to handle it via the LED API, why is a new binding needed?

The main reason is to follow the DTS bindings in Linux. As I understood it, Zephyr want to be compatible with Linux devicetree (as long as it makes sense). 

These are the places in the Linux kernel that inspired this binding (in the same order as the PR description):
1. `led-backlight` binding ([src](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/leds/backlight/led-backlight.yaml#L4))
2. `panel-common` has a `backlight`property ([src](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/display/panel/panel-common.yaml#L146))
3. `librem5` board has a `led-backlight` compatible ([src](https://elixir.bootlin.com/linux/latest/source/arch/arm64/boot/dts/freescale/imx8mq-librem5.dtsi#L18))

henrikbrixandersen commented:
I don't see the need for a dedicated binding for this, since we don't intend to introduce a new driver for LCD backlights. A simple phandle property pointing to the LED dts node should suffice.

What do others think? @galak, @mbolivar-nordic?

jfischer-no commented:
> I don't see the need for a dedicated binding for this, since we don't intend to introduce a new driver for LCD backlights. A simple phandle property pointing to the LED dts node should suffice.

I agree.

caspermeijn commented:
@henrikbrixandersen @jfischer-no I changed the code to be a single phandle from the display to the led. 

To be clear, this version is the same as the Linux bindings anymore.

jfischer-no commented on dts/bindings/display/display-controller.yaml:
```diff
@@ -18,3 +18,11 @@ properties:
     label:
       required: true
+
+    backlight-led:
```
Generally, a display controller is not a display (even if for smaller displays like ST7789V one can say it is a unit). A display (panel) can have a backlight, what does not necessarily have to be a LED, or none at all (OLED, EPD). So this does not fit here. Probably you can put it in ST7789V bindings.

caspermeijn replied:
You could use the same arguments for height and width, but lets not discuss that in this issue. 

I moved the property to the ST7789V binding.

caspermeijn commented on dts/bindings/display/display-controller.yaml:
```diff
@@ -18,3 +18,11 @@ properties:
+    backlight-led:
+      type: phandle
+      required: false
+      description:
```
Fixed

henrikbrixandersen reviewed (dismissed):
LGTM

caspermeijn commented:
I don't understand why the build fails. I just rebased on main and the build fails again with the message:
```
 FileNotFoundError: [Errno 2] No such file or directory: '_test_plan_partial.csv'
```

mbolivar-nordic commented:
No idea, but you might want to try again

caspermeijn commented:
> No idea, but you might want to try again

It fails again :-/

caspermeijn commented:
> > > No idea, but you might want to try again
> > 
> > 
> > It fails again :-/
> 
> Really sorry about that. Now that the merge window for v3.1 is about to reopen, I suggest rebasing one more time and seeing what happens. If it fails again and I can't figure out why, I will pull in the infrastructure maintainers.

@mbolivar-nordic I tried rebasing twice, without progress. I also see this problem in other PR for pinetime board.

mbolivar-nordic commented:
@caspermeijn https://github.com/zephyrproject-rtos/zephyr/pull/43281 is merged and should be a fix -- please try rebasing again, and thanks for your patience.

henrikbrixandersen reviewed (commented):
Shouldn't there be a corresponding driver change adding support for the backlight LED?

caspermeijn commented:
> Shouldn't there be a corresponding driver change adding support for the backlight LED?

I primarily want to describe this hardware relation, so that I can enable the backlight in my application. I want to extend the ST7789V to use the led api to enable the backlight, but it seems that it is not yet possible to follow the phandle to address a led (https://github.com/zephyrproject-rtos/zephyr/issues/42015).

caspermeijn committed:
boards: arm: pinetime: Enable LED_GPIO driver for DISPLAY

Enable the led gpio driver when the display is requested. This way the
backlight can be turned on at startup.

Signed-off-by: Casper Meijn <casper@meijn.net>

caspermeijn commented:
With a macro for retrieving the led number, it is now possible to extend the st7789v driver with powering the backlight. Therefore I added that to the PR.

mbolivar-nordic commented on drivers/display/display_st7789v.c:
```diff
@@ -413,6 +434,10 @@ static const struct st7789v_config st7789v_config = {
 	.reset_gpio = GPIO_DT_SPEC_INST_GET(0, reset_gpios),
 #endif
+#if DT_INST_NODE_HAS_PROP(0, backlight_led)
+	.backlight_dev = DEVICE_DT_GET_OR_NULL(DT_PARENT(DT_INST_PHANDLE(0, backlight_led))),
+	.backlight_led = DT_NODE_CHILD_IDX(DT_INST_PHANDLE(0, backlight_led)),
```
I will leave this change to @Mani-Sadhasivam to ACK, because I'm not sure it's correct. It seems like this might be a little bit hard-coded to the gpio-leds compatible. Are there any other types of LED in devicetree that you might need to handle differently?

Mani-Sadhasivam replied:
I think this is fine IMO. Only the `gpio-leds` are special as they are referenced by different applications like LCD, Keyboard etc... for backlight purposes. The LED controllers mostly have a dedicated application controlling them as a whole so the knowledge of the pattern is obvious.

Moreover, the usage of a dedicated property allows us to follow what Linux has been doing for decades.

mbolivar-nordic commented:
OK! We have all the approvals we need. @caspermeijn unfortunately there are some conflicts with your branch -- can you please rebase again? Then @Mani-Sadhasivam, please re-approve, and I will merge.

caspermeijn commented:
@mbolivar-nordic @Mani-Sadhasivam Rebased on main

Mani-Sadhasivam reviewed (approved):
LED changes LGTM!

jfischer-no commented on drivers/display/display_st7789v.c:
```diff
@@ -206,7 +209,13 @@ static void *st7789v_get_framebuffer(const struct device *dev)
+	const struct st7789v_config *config = dev->config;
+
+	if (config->backlight_dev) {
+		return led_set_brightness(config->backlight_dev, config->backlight_led, brightness);
```
It seems to me that here board specific things are hacked into the driver. I am blocking it for now.

caspermeijn replied:
@jfischer-no Is there a way to progress this patch? My goal is to have a backlight brightness interface that is not board specific. To me the `display_set_brightness` seems to be just that.

ljunquera commented:
Was this resolved?
